doc/sphinx/proof-engine/tactics.rst

```diff
@@ -466,52 +466,87 @@ Examples:
-(see :ref:`ImplicitArguments`) or coercions (see :ref:`Coercions`) are
-counted.
-
-If a minus sign is given between ``at`` and the list of occurrences, it
-negates the condition so that the clause denotes all the occurrences
-except the ones explicitly mentioned after the minus sign.
-
-As an exception to the left-to-right order, the occurrences in
-the return subexpression of a match are considered *before* the
-occurrences in the matched term.
-
-In the second case, the ``*`` on the left of ``|-`` means that all occurrences
-of term are selected in every hypothesis.
-
-In the first and second case, if ``*`` is mentioned on the right of ``|-``, the
-occurrences of the conclusion of the goal have to be selected. If some numbers
-are given, then only the occurrences denoted by these numbers are selected. If
-no numbers are given, all occurrences of :token:`term` in the goal are selected.
-
-Finally, the last notation is an abbreviation for ``* |- *``. Note also
-that ``|-`` is optional in the first case when no ``*`` is given.
-
-Here are some tactics that understand occurrence clauses: :tacn:`set`,
-:tacn:`remember`, :tacn:`induction`, :tacn:`destruct`.
+   .. insertprodn occurrences concl_occs
+
+   .. prodn::
+      occurrences ::= at @occs_nums
+      | in @goal_occurrences
+      occs_nums ::= {? - } {+ {| @natural | @ident } }
+      goal_occurrences ::= {*, @hyp_occs } {? %|- {? @concl_occs } }
+      | * %|- {? @concl_occs }
+      | * {? at @occs_nums }
+      hyp_occs ::= @hypident {? at @occs_nums }
+      hypident ::= @ident
+      | ( type of @ident )
+      | ( value of @ident )
+      concl_occs ::= * {? at @occs_nums }
+
+   :n:`@occurrences`
+     The first form of :token:`occurrences` selects occurrences in
+     the conclusion of the goal.  The second form can select occurrences
+     in the goal conclusion and in one or more hypotheses.
+
+   :n:`{? - } {+ {| @natural | @ident } }`
+     Selects the specified occurrences within a single goal or hypothesis.
+     Occurrences are numbered from left to right starting with 1 when the
+     goal is printed with the :flag:`Printing All`.  (In particular, occurrences
+     in implicit arguments (see :ref:`ImplicitArguments`) and
+     coercions (see :ref:`Coercions`) are counted but not shown by default.)
+
+     Specifying `-` includes all occurrences *except* the ones listed.
+
+   :n:`{*, @hyp_occs } {? %|- {? @concl_occs } }`
+     Selects occurrences in the specified hypotheses and the
+     specified occurrences in the conclusion.
+
+     .. todo PR: if above is empty it's useless, right?
+
+   :n:`* %|- {? @concl_occs }`
+     Selects all occurrences in all hypotheses and the
+     specified occurrences in the conclusion.
+
+   :n:`@goal_occurrences ::= * {? at @occs_nums }`
+     Selects all occurrences in all hypotheses and in the conclusion.
+
+     .. todo PR: above is likely wrong.  If * selects everything hyp+concl, the
+        occs_nums don't make sense
```

Comment:
Indeed. According to my testing, `* at ...` is a shorthand for `* |- * at ...`, i.e. it selects all the occurrences in the hypotheses and only the specified occurrences in the goal. Though, I think this syntax is counter-intuitive and should maybe be deprecated (I mean the shorthand `*` would be kept but without any optional `at`).

PS: during these tests, I also discovered that the documented syntax of `occs_nums` does not seem to be right: one can repeat the minus sign (and it has the same effect as if it wasn't repeated).

```coq
Parameter b : nat.
Definition a := b.
Goal a + a = 0 -> a = a /\ a = 0.
intro.
(*
H : a + a = 0
________________
a = a /\ a = 0
*)
change a with b in * at -1 -2.
```

However, one cannot insert a minus sign after a non-negative integer (as in `at 1 -2`).
This might be a parsing bug and is certainly not worth being documented.

Reply:
I would make these changes:
- remove the `at` clause after `*` in `goal_occurrences` as you suggest
- make concl_occs accept '*' OR 'at occs_nums`
We could make these optional and  deprecated.

`occs_nums` is defined this way in the mlgs, which explains why `in 1 -2` is not valid.  It should have `nat_or_var` instead of `int_or_var`.  I was going to put in a PR, I guess it's too late for 8.13.  You think it's worth deprecating the negative numbers since it's a bit strange to use them?

```
occs_nums: [
| LIST1 nat_or_var
| "-" nat_or_var LIST0 int_or_var
]
```

Reply:
For the second bullet-point, I don't view this change as necessary / useful (there is only one goal but it is unnamed and thus we select it with `*` instead of a name).

Yes to the rest (I see that you've already opened #13403 for the last point).

Reply:
Until the first bullet point is actually fixed, maybe the optional `at` clause following a `*` without a `|-` should be erased from the manual?

Reply:
First bullet point: Ah, I read the code.  `*` by itself means all occurrences in the conclusion.  If you have `* at @occs_num` then only the listed occurrences are selected.  So my todo was incorrect and it should stay as is.  See `in_clause` and `clause_dft_concl` in `g_tactic.mlg`.

For the second bullet point: `*` generally means "all".  So to me, having `* at @occs_nums` in which the `occs_nums` give the subset seems inconsistent.